File: words.py
class Words:
    def __init__(self, file_name): #Constructor takes file name as argument
        self.words = [] #Define list
        with open(file_name) as word_dictionary: #Open the file for reading
            for word in word_dictionary: #Read until file ends
                self.words.append(word.strip('\n')) #Remove trailing white spaces and append to the list
    
    def is_word(self, word): #Check whether word is present in the list or not
        return word in self.words #Returns true or false depending on whether the words are present in the list or not

    def get_size(self): #Get the size of the word
        return len(self.words) #Returns the size of the word

    def get_replace_corrections(self, misspelled_word): #Returns the potential list of corrections for the word
        suggestions = [] #Creates a list to store the suggestions
        for w in self.words: #Loop through all the words found from the file
            # only consider words that have the same length as misspelled_word
            if len(w) == len(misspelled_word):
                # try replacing one letter at a time and see if we can transform the misspelled word into word w
                for index in range(0, len(w)): #Loop through every position of the word
                    correction = misspelled_word[0:index]+w[index] #Construct a new string called correction by replacing the letter at that position
                    if index < len(w)-1: #Check if the index is less than the length of the word
                        correction+=misspelled_word[index+1:] #Correction transformation
                    # test if replaced letter transforms our word to w    
                    if correction == w: #If transformed word matches the word
                        suggestions.append(w) #Append the transformed word to suggestions list and exit the loop
                        break
        return suggestions #Return the suggested words
        
    def get_delete_corrections(self, misspelled_word): #Check if transformation is possible by deletion
        suggestions = [] #Initialize suggestions list
        for w in self.words: #Loop through the words
            # only consider words that are one character shorter than misspelled_word
            if len(w) == len(misspelled_word)-1:
                # try deleting one character from misspelled word at a time and see if it transforms to word w
                if self.can_ransform_with_delete(misspelled_word, w):
                    suggestions.append(w)
        return suggestions
    
    def get_add_corrections(self, misspelled_word):
        suggestions = []
        for w in self.words:
            # only consider words that are one character longer than misspelled word
            if len(w) == len(misspelled_word)+1:
                if self.can_ransform_with_delete(w, misspelled_word):
                    suggestions.append(w)
        return suggestions

    def can_ransform_with_delete(self, source, dest): #Check whether source and destionation word match
        transformed = False #Assume false by default
        for index in range(0, len(source)): #Loop through every letter of the word
            correction = source[0:index] #Create a new correction string from 0 to the end of the character string
            if index < len(source)-1: #Append the string from the omitted letter till the end
                correction+=source[index+1:]
            # test if deleted letter transforms our word to w
            if correction == dest: #Check if the transofrmed letter is equal to the destination then set to true
                transformed = True
                break
        return transformed #Return the transformed word

File: Words.java
import java.util.Scanner;
import java.util.HashSet;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;

public class Words{
    private HashSet<String> words;
    public Words(String fileName){
        words = new HashSet<String>();
        String line = "";
        try{
            Scanner scanner = new Scanner(new FileInputStream(new File(fileName)));
            while (scanner.hasNextLine()){
                line = scanner.nextLine();
                words.add(line);
            }
        }
        catch (IOException error){
            System.out.println("Failed to construct Words " + error.getMessage());
        }
    }

    public boolean isWord(String word){
        boolean result = true;
        if (!words.contains(word)) {
            result = false;
        }
        return result;
    }

    public int getSize(){
        return words.size();
    }

    public String[] getReplaceCorrections(String misspelledWord){
        ArrayList<String> suggestions = new ArrayList<String>();
        for (String word: this.words){
            if (word.length() == misspelledWord.length()){
                for (int i = 0; i < misspelledWord.length(); i++){
                    String correction = misspelledWord.substring(0, i)+word.charAt(i);
                    if ( i < misspelledWord.length() - 1){
                        correction+=misspelledWord.substring(i+1);
                    }
                    if (correction.equals(word)){
                        suggestions.add(correction);
                        break;
                    }
                }

            }
        }
        return arrayListToArray(suggestions);
    }

    public String[] getDeleteCorrections(String misspelledWord){
        ArrayList<String> suggestions = new ArrayList<String>();
        for (String word: this.words){
            if (word.length() == misspelledWord.length()-1){
                if (canTransformWithDelete(misspelledWord, word)){
                    suggestions.add(word);
                }
            }
        }
        return arrayListToArray(suggestions);
    }

    public String[] getAddCorrections(String misspelledWord){
        ArrayList<String> suggestions = new ArrayList<String>();
        for (String word: this.words){
            if (word.length() == misspelledWord.length()+1){
                if (canTransformWithDelete(word, misspelledWord)){
                    suggestions.add(word);
                }
            }
        }
        return arrayListToArray(suggestions);
    }

    private boolean canTransformWithDelete(String from, String to){
        boolean transformed = false;
        for (int i = 0; i < from.length(); i++){
            String correction = from.substring(0, i);
            if ( i < from.length() - 1){
                correction+=from.substring(i+1);
            }
            if (correction.equals(to)){
                transformed = true;
                break;
            }
        }
        return transformed;
    }

    private String[] arrayListToArray(ArrayList<String> suggestions){
        String []result = new String[suggestions.size()];
        int index = 0;
        for (String w: suggestions){
            result[index] = w;
            index++;
        }
        return result;
    }
}